certificates.go

```diff
@@ -67,10 +73,106 @@ func (cert Certificate) Empty() bool {
+	// (code borrowed from our acme package)
+	if selectedTime.IsZero() &&
+		(!ari.SuggestedWindow.Start.IsZero() && !ari.SuggestedWindow.End.IsZero()) {
+		start, end := ari.SuggestedWindow.Start.Unix()+1, ari.SuggestedWindow.End.Unix()
+		selectedTime = time.Unix(rand.Int63n(end-start)+start, 0).UTC()
+		logger.Warn("no renewal time had been selected with ARI; chose an ephemeral one for now",
+			zap.Time("ephemeral_selected_time", selectedTime))
+	}
+
+	// if a renewal time has been selected, start with that
+	if !selectedTime.IsZero() {
+		// ARI spec recommends an algorithm that renews after the randomly-selected
+		// time OR just before it if the next waking time would be after it; this
+		// cutoff can actually be before the start of the renewal window, but the spec
+		// author says that's OK: https://github.com/aarongable/draft-acme-ari/issues/71
+		cutoff := ari.SelectedTime.Add(-cfg.certCache.options.RenewCheckInterval)
+		if time.Now().After(cutoff) {
+			logger.Info("certificate needs renewal based on ARI window",
+				zap.Time("selected_time", selectedTime),
+				zap.Time("renewal_cutoff", cutoff))
+			return true
+		}
+
+		// according to ARI, we are not ready to renew; however, we do not rely solely on
+		// ARI calculations... what if there is a bug in our implementation, or in the
+		// server's, or the stored metadata? for redundancy, give credence to the expiration
+		// date; ignore ARI if we are past a "dangerously close" limit, to avoid any
+		// possibility of a bug in ARI compromising a site's uptime: we should always always
+		// always give heed to actual validity period
+		if currentlyInRenewalWindow(leaf.NotBefore, expiration, 1.0/20.0) {
+			logger.Warn("certificate is in emergency renewal window; superceding ARI",
+				zap.Duration("remaining", time.Until(expiration)),
+				zap.Time("renewal_cutoff", cutoff))
+			return true
+		}
+
+	}
+
+	// the normal check, in the absence of ARI, is to determine if we're near enough (or past)
+	// the expiration date based on the configured remaining:lifetime ratio
+	if currentlyInRenewalWindow(leaf.NotBefore, expiration, cfg.RenewalWindowRatio) {
+		logger.Info("certificate is in configured renewal window based on expiration date",
+			zap.Duration("remaining", time.Until(expiration)))
+		return true
+	}
+
+	// finally, if the certificate is expiring imminently, always attempt a renewal;
+	// we check both a (very low) lifetime ratio and also a strict difference between
+	// the time until expiration and the interval at which we run the standard maintenance
+	// routine to check for renewals, to accommodate both exceptionally long and short
+	// cert lifetimes
+	if currentlyInRenewalWindow(leaf.NotBefore, expiration, 1.0/50.0) ||
+		time.Until(expiration) < cfg.certCache.options.RenewCheckInterval*5 {
+		logger.Warn("certificate is in emergency renewal window; expiration imminent",
+			zap.Duration("remaining", time.Until(expiration)))
+		return true
+	}
+
+	return false
```

Comment:
The comments look sensible. Maybe some unit test of thie `certNeedsRenewal` function would be beneficial as well? (to ensure that any change does not break expected behavior)